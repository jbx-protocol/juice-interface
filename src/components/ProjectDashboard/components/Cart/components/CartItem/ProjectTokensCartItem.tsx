import { InformationCircleIcon } from '@heroicons/react/24/outline'
import { t } from '@lingui/macro'
import { Tooltip } from 'antd'
import { ProjectHeaderLogo } from 'components/ProjectDashboard/components/ProjectHeader/components/ProjectHeaderLogo'
import { useProjectPaymentTokens } from 'components/ProjectDashboard/hooks/useProjectPaymentTokens'
import { CartItem } from './CartItem'
import { CartItemBadge } from './CartItemBadge'

export const ProjectTokensCartItem = () => {
  const { receivedTickets } = useProjectPaymentTokens()

  return (
    <CartItem
      title={
        <div className="flex flex-col gap-1">
          <span className="flex items-center gap-2">
            <span>Project tokens</span>
            <Tooltip
              title={t`Payments to Juicebox projects automatically mint tokens`}
            >
              <InformationCircleIcon className="h-4 w-4 text-grey-400 dark:text-slate-200" />
            </Tooltip>
            <CartItemBadge>Token</CartItemBadge>
          </span>
          <span>{receivedTickets}</span>
        </div>
      }
      icon={<ProjectHeaderLogo className="h-14 w-14 rounded-full" />}
      price={null}
    />
  )
}
