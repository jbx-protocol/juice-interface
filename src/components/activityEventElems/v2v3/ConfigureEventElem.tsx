import { t, Trans } from '@lingui/macro'
import { MinimalCollapse } from 'components/MinimalCollapse'
import RichNote from 'components/RichNote'
import { BigNumber } from 'ethers'
import {
  V2V3FundingCycle,
  V2V3FundingCycleMetadata,
} from 'models/v2v3/fundingCycle'

import FundingCycleDetails from 'components/v2v3/V2V3Project/V2V3FundingCycleSection/FundingCycleDetails'
import { PV_V2 } from 'constants/pv'
import { V2V3ProjectContext } from 'contexts/v2v3/Project/V2V3ProjectContext'
import { ProjectEventsQuery } from 'generated/graphql'
import useProjectDistributionLimit from 'hooks/v2v3/contractReader/useProjectDistributionLimit'
import { useContext } from 'react'
import { ActivityEvent } from '../ActivityElement'

export default function ConfigureEventElem({
  event,
  withProjectLink,
}: {
  event: ProjectEventsQuery['projectEvents'][0]['configureEvent']
  withProjectLink?: boolean
}) {
  if (!event) return null

  const { primaryETHTerminal } = useContext(V2V3ProjectContext)

  const { data: distributionLimit } = useProjectDistributionLimit({
    projectId: event.projectId,
    terminal: primaryETHTerminal,
    configuration: event.configuration?.toString(),
  })

  const fundingCycle: Partial<V2V3FundingCycle> = {
    duration: BigNumber.from(event.duration),
    weight: BigNumber.from(event.weight),
    discountRate: BigNumber.from(event.discountRate),
    ballot: event.ballot,
    start: event.mustStartAtOrAfter
      ? BigNumber.from(event.mustStartAtOrAfter)
      : undefined,
  }

  const fundingCycleMetadata: Partial<V2V3FundingCycleMetadata> = {
    reservedRate: BigNumber.from(event.reservedRate),
    redemptionRate: BigNumber.from(event.redemptionRate),
    global: {
      pauseTransfers: event.transfersPaused,
      allowSetController: event.setControllerAllowed,
      allowSetTerminals: event.setTerminalsAllowed,
    },
    allowControllerMigration: event.controllerMigrationAllowed,
    allowTerminalMigration: event.terminalMigrationAllowed,
    pausePay: event.pausePay,
    pauseRedeem: event.redeemPaused,
    pauseBurn: event.burnPaused,
    pauseDistributions: event.distributionsPaused,
    useTotalOverflowForRedemptions: event.useTotalOverflowForRedemptions,
    holdFees: event.shouldHoldFees,
    dataSource: event.dataSource,
    useDataSourceForPay: event.useDataSourceForPay,
    useDataSourceForRedeem: event.useDataSourceForRedeem,
    allowMinting: event.mintingAllowed,
  }

  return (
    <ActivityEvent
      event={event}
      header={t`Edited cycle`}
      withProjectLink={withProjectLink}
      pv={PV_V2}
      subject={
        <MinimalCollapse
          header={
            <span className="font-normal">
              <Trans>Details</Trans>
            </span>
          }
        >
          <FundingCycleDetails
            fundingCycle={fundingCycle as V2V3FundingCycle}
            fundingCycleMetadata={
              fundingCycleMetadata as V2V3FundingCycleMetadata
            }
            distributionLimit={distributionLimit?.[0]}
            distributionLimitCurrency={distributionLimit?.[1]}
            mintRateZeroAsUnchanged
          />
          {event.memo ? <RichNote className="mt-3" note={event.memo} /> : null}
        </MinimalCollapse>
      }
    />
  )
}
