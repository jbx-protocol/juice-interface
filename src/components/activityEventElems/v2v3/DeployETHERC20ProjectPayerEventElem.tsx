import { t, Trans } from '@lingui/macro'
import EthereumAddress from 'components/EthereumAddress'
import RichNote from 'components/RichNote'
import { ProjectEventsQuery } from 'generated/graphql'

import { PV_V2 } from 'constants/pv'
import { ActivityEvent } from '../ActivityElement'

export default function DeployETHERC20ProjectPayerEventElem({
  event,
  withProjectLink,
}: {
  event:
    | ProjectEventsQuery['projectEvents'][0]['deployETHERC20ProjectPayerEvent']
  withProjectLink?: boolean
}) {
  if (!event) return null

  return (
    <ActivityEvent
      event={event}
      header={t`Deployed a project payer address`}
      withProjectLink={withProjectLink}
      pv={PV_V2}
      subject={
        <Trans>
          from <EthereumAddress address={event.from} />
        </Trans>
      }
      extra={
        <div>
          <Trans>
            Address: <EthereumAddress address={event.address} />
          </Trans>
          {event.memo && (
            <div className="mt-2">
              <RichNote note={event.memo} />
            </div>
          )}
        </div>
      }
    />
  )
}
