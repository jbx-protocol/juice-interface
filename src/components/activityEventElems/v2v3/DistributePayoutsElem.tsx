import { t } from '@lingui/macro'
import EthereumAddress from 'components/EthereumAddress'
import { JuiceboxAccountLink } from 'components/JuiceboxAccountLink'
import ETHAmount from 'components/currency/ETHAmount'
import V2V3ProjectHandleLink from 'components/v2v3/shared/V2V3ProjectHandleLink'
import { PV_V2 } from 'constants/pv'
import {
  ProjectEventsQuery,
  useSplitDistributionsForDistributePayoutsEventQuery,
} from 'generated/graphql'
import { client } from 'lib/apollo/client'

import { ActivityEvent } from '../ActivityElement'

export default function DistributePayoutsElem({
  event,
  withProjectLink,
}: {
  event: ProjectEventsQuery['projectEvents'][0]['distributePayoutsEvent']
  withProjectLink?: boolean
}) {
  // Load individual DistributeToPayoutSplit events, emitted by internal transactions of the DistributeReservedPayouts transaction
  const { data } = useSplitDistributionsForDistributePayoutsEventQuery({
    client,
    variables: {
      distributePayoutsEvent: event?.id,
    },
  })

  if (!event) return null

  return (
    <ActivityEvent
      event={event}
      header={t`Sent payouts`}
      withProjectLink={withProjectLink}
      pv={PV_V2}
      subject={
        data?.distributeToPayoutSplitEvents.length ? (
          <span className="font-heading text-lg">
            <ETHAmount amount={event.distributedAmount} />
          </span>
        ) : null
      }
      extra={
        <div>
          {data?.distributeToPayoutSplitEvents.map(e => (
            <div
              key={e.id}
              className="flex items-baseline justify-between text-sm"
            >
              <div className="font-medium">
                {e.splitProjectId ? (
                  <V2V3ProjectHandleLink
                    className="text-grey-900 dark:text-slate-100"
                    projectId={e.splitProjectId}
                  />
                ) : (
                  <JuiceboxAccountLink
                    className="text-grey-900 dark:text-slate-100"
                    address={e.beneficiary}
                  />
                )}
                :
              </div>

              <div className="text-secondary">
                <ETHAmount amount={e.amount} />
              </div>
            </div>
          ))}

          {event.beneficiaryDistributionAmount?.gt(0) && (
            <div className="flex items-baseline justify-between">
              <div>
                <EthereumAddress
                  className="text-grey-900 dark:text-slate-100"
                  address={event.beneficiary}
                />
                :
              </div>
              <div className="text-sm text-grey-500 dark:text-grey-300">
                <ETHAmount amount={event.beneficiaryDistributionAmount} />
              </div>
            </div>
          )}
        </div>
      }
    />
  )
}
