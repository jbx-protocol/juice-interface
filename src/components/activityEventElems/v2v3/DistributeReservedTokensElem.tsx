import { Trans } from '@lingui/macro'
import EthereumAddress from 'components/EthereumAddress'
import { JuiceboxAccountLink } from 'components/JuiceboxAccountLink'
import { PV_V2 } from 'constants/pv'
import { V1ProjectContext } from 'contexts/v1/Project/V1ProjectContext'
import {
  ProjectEventsQuery,
  useSplitDistributionsForDistributeReservedTokensEventQuery,
} from 'generated/graphql'
import { client } from 'lib/apollo/client'
import { useContext } from 'react'
import { formatWad, fromWad } from 'utils/format/formatNumber'
import { tokenSymbolText } from 'utils/tokenSymbolText'
import { ActivityEvent } from '../ActivityElement'

export default function DistributeReservedTokensEventElem({
  event,
  withProjectLink,
}: {
  event: ProjectEventsQuery['projectEvents'][0]['distributeReservedTokensEvent']
  withProjectLink?: boolean
}) {
  const { tokenSymbol } = useContext(V1ProjectContext)

  // Load individual DistributeToReservedTokenSplit events, emitted by internal transactions of the DistributeReservedTokens transaction
  const { data } = useSplitDistributionsForDistributeReservedTokensEventQuery({
    client,
    variables: {
      distributeReservedTokensEvent: event?.id,
    },
  })

  if (!event) return null

  return (
    <ActivityEvent
      event={event}
      withProjectLink={withProjectLink}
      pv={PV_V2}
      header={
        <Trans>
          Sent reserved{' '}
          {tokenSymbolText({
            tokenSymbol,
            capitalize: false,
            plural: true,
          })}
        </Trans>
      }
      subject={
        <div className="font-heading text-lg">
          {formatWad(event.tokenCount, { precision: 0 })}{' '}
          {tokenSymbolText({
            tokenSymbol,
            capitalize: false,
            plural: parseInt(fromWad(event.tokenCount) || '0') !== 1,
          })}
        </div>
      }
      extra={
        <div>
          {data?.distributeToReservedTokenSplitEvents.map(e => (
            <div key={e.id} className="flex items-baseline justify-between">
              <div>
                <JuiceboxAccountLink
                  className="text-grey-900 dark:text-slate-100"
                  address={e.beneficiary}
                />
              </div>

              <div className="text-sm text-grey-500 dark:text-grey-300">
                {formatWad(e.tokenCount, { precision: 0 })}{' '}
                {tokenSymbolText({
                  tokenSymbol,
                  capitalize: false,
                  plural: true,
                })}
              </div>
            </div>
          ))}

          {event.beneficiaryTokenCount?.gt(0) && (
            <div className="flex items-baseline justify-between">
              <div>
                <EthereumAddress
                  className="text-grey-900 dark:text-slate-100"
                  address={event.beneficiary}
                />
                :
              </div>
              <div className="text-sm text-grey-500 dark:text-grey-300">
                {formatWad(event.beneficiaryTokenCount, {
                  precision: 0,
                })}
              </div>
            </div>
          )}
        </div>
      }
    />
  )
}
