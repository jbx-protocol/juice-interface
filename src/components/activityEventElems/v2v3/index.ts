export * from './ConfigureEventElem'
export * from './DeployETHERC20ProjectPayerEventElem'
export * from './DistributePayoutsElem'
export * from './DistributeReservedTokensElem'
export * from './SetFundAccessConstraintsEventElem'
