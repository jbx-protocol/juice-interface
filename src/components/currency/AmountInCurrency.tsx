import { CurrencyName } from 'constants/currency'
import { BigNumber } from 'ethers'
import ETHAmount from './ETHAmount'
import USDAmount from './USDAmount'

/**
 *
 * @returns amount in ETH by default. Specified currency otherwise
 */
export function AmountInCurrency({
  currency,
  amount,
  ...props
}: {
  amount: BigNumber | undefined
  currency?: CurrencyName
}) {
  if (currency === 'USD') return <USDAmount amount={amount} {...props} />
  return <ETHAmount amount={amount} {...props} />
}
