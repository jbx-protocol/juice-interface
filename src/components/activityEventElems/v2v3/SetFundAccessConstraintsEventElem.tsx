import { t } from '@lingui/macro'
import { AmountInCurrency } from 'components/currency/AmountInCurrency'
import { ProjectEventsQuery } from 'generated/graphql'
import { V2V3CurrencyOption } from 'models/v2v3/currencyOption'
import { V2V3CurrencyName } from 'utils/v2v3/currency'
import { MAX_DISTRIBUTION_LIMIT } from 'utils/v2v3/math'

import { PV_V2 } from 'constants/pv'
import { ActivityEvent } from '../ActivityElement'

export default function SetFundAccessConstraintsEventElem({
  event,
  withProjectLink,
}: {
  event: ProjectEventsQuery['projectEvents'][0]['setFundAccessConstraintsEvent']
  withProjectLink?: boolean
}) {
  if (!event) return null

  return (
    <ActivityEvent
      event={event}
      withProjectLink={withProjectLink}
      pv={PV_V2}
      header={t`Edited payout`}
      subject={
        <div className="font-heading text-lg">
          {event.distributionLimit.eq(MAX_DISTRIBUTION_LIMIT) ? (
            t`Unlimited`
          ) : (
            <AmountInCurrency
              amount={event.distributionLimit}
              currency={V2V3CurrencyName(
                event.distributionLimitCurrency as V2V3CurrencyOption,
              )}
            />
          )}
        </div>
      }
    />
  )
}
