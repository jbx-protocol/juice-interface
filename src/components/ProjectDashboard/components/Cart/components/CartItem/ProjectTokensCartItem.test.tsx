/**
 * @jest-environment jsdom
 */
import { render } from '@testing-library/react'
import {
  useProjectCart,
  useTokensPerEth,
} from 'components/ProjectDashboard/hooks'
import { V2V3_CURRENCY_ETH } from 'utils/v2v3/currency'
import { ProjectTokensCartItem } from './ProjectTokensCartItem'

jest.mock('components/ProjectDashboard/hooks')
jest.mock(
  'components/ProjectDashboard/components/ProjectHeader/components/ProjectHeaderLogo',
  () => ({
    ProjectHeaderLogo: () => <div>ProjectHeaderLogo</div>,
  }),
)

describe('ProjectTokensCartItem', () => {
  const DefaultUseProjectCartMock = {
    dispatch: jest.fn(),
    payAmount: {
      amount: 100,
      currency: V2V3_CURRENCY_ETH,
    },
    expanded: false,
    visible: true,
  }

  const UseTokensPerEthMock = {
    receivedTickets: '100',
  }
  beforeEach(() => {
    ;(useProjectCart as jest.Mock).mockReturnValue(DefaultUseProjectCartMock)
    ;(useTokensPerEth as jest.Mock).mockReturnValue(UseTokensPerEthMock)
  })

  it('renders', () => {
    const { container } = render(<ProjectTokensCartItem />)
    expect(container).toMatchSnapshot()
  })
})
