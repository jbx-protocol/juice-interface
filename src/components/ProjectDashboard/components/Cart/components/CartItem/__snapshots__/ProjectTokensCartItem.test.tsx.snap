// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectTokensCartItem renders 1`] = `
<div>
  <div
    class="flex min-w-0 items-center justify-between gap-4 border-b border-grey-200 py-5 dark:border-slate-500"
  >
    <div
      class="flex min-w-0 items-center"
    >
      <div>
        ProjectHeaderLogo
      </div>
      <span
        class="ml-3 min-w-0 font-medium dark:text-slate-50"
      >
        <div
          class="flex flex-col gap-1"
        >
          <span
            class="flex items-center gap-2"
          >
            <span>
              Project tokens
            </span>
            <svg
              aria-hidden="true"
              class="h-4 w-4 text-grey-400 dark:text-slate-200"
              fill="none"
              stroke="currentColor"
              stroke-width="1.5"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M11.25 11.25l.041-.02a.75.75 0 011.063.852l-.708 2.836a.75.75 0 001.063.853l.041-.021M21 12a9 9 0 11-18 0 9 9 0 0118 0zm-9-3.75h.008v.008H12V8.25z"
                stroke-linecap="round"
                stroke-linejoin="round"
              />
            </svg>
            <span
              class="whitespace-nowrap rounded-2xl bg-grey-100 py-0.5 px-2 text-xs dark:bg-slate-500 dark:text-slate-100"
            >
              Token
            </span>
          </span>
          <span>
            100
          </span>
        </div>
      </span>
    </div>
    <div
      class="flex items-center"
    >
      <span
        class="mr-4 text-sm font-medium"
      >
        -
      </span>
    </div>
  </div>
</div>
`;
